src/arch/x86/object/vcpu.c

```diff
@@ -517,14 +549,124 @@ static exception_t invokeVCPUWriteRegisters(vcpu_t *vcpu, word_t *buffer)
 
 static exception_t decodeVCPUWriteRegisters(cap_t cap, word_t length, word_t *buffer)
 {
+#ifdef CONFIG_ARCH_X86_64
+    if (length < 15) {
+#else
     if (length < 7) {
+#endif /* CONFIG_ARCH_X86_64 */
         userError("VCPU WriteRegisters: Truncated message.");
         current_syscall_error.type = seL4_TruncatedMessage;
         return EXCEPTION_SYSCALL_ERROR;
     }
     return invokeVCPUWriteRegisters(VCPU_PTR(cap_vcpu_cap_get_capVCPUPtr(cap)), buffer);
 }
 
+#ifdef CONFIG_ARCH_X86_64
+static exception_t invokeWriteMSR(vcpu_t *vcpu, word_t *buffer, word_t field, word_t value)
```

Comment:
Is it possible that the read and write MSR methods are being invoked by another thread running on core1 which the VCPU is running on core0?

Reply:
These values are tracked per-vcpu, and are saved and restored as such. Since `wrmsr` and `rdmsr` are trapped into the hypervisor code anyways, I don't think its possible to mess with a guest in this way, since the MSR values returned to the guest are values stored in the vcpu structure.

```
#ifdef CONFIG_ARCH_X86_64
        case MSR_EFER:
            data = vmm_vmcs_read(vcpu->guest_vcpu, VMX_GUEST_EFER);
            break;

        case MSR_STAR:
        case MSR_LSTAR:
        case MSR_CSTAR:
        case MSR_SYSCALL_MASK:
            data = vmm_msr_read(vcpu->guest_vcpu, msr_no);
            break;
#endif
```